### ISSUE TITLE
Possibility to display line endings in text area

### ISSUE THREAD

**Issue by tonka3000** [Microsoft/vscode#12223]
- VSCode Version: 1.5
- OS Version: Win 7 x64

At the moment there is only the little statusbar menu which display/change the line ending if the actual file. But sometimes it would be great to see the line endings directly in every line (see atom f.e.) especially when there are mixed line endings (not good, but this happens from time to time).

Greetings
Tonka

**Comment by jlarcombe** [comment]
Same thing as #24625

**Comment by masaeedu** [comment]
@alexandrudima Any progress on this front? Keep having to crack open notepad++ or `cat -e` to see newlines, which is pretty annoying.

**Comment by fabulousduck** [comment]
Any update on this ?
And it seems that this issue is looping since there are 2 issues that refer to each other (#24625 <> #12223)

**Comment by sajidali** [comment]
@alexandrudima Any update on this ?

**Comment by sajidali** [comment]
Add "editor.renderWhitespace": "all" to settings.json
It renders most whitespaces, but not newline characters.

It would be very helpful to render the newline characters, with distinct renderings for CR vs LF. This is especially helpful if you have a mixed line ending file and want to see where the issues are.
@bpasero Is there any chance, someone picks this up and completes this?

**Comment by djtrustgod** [comment]
@sajidali I agree.  I keep having to copy my code into NotePad ++ to see special characters like new lines.  I would like to see `editor.renderControlCharacters` so it behaves.

**Comment by danPadric** [comment]
Any movement on this case for rendering the line endings?

**Comment by adamralph** [comment]
This issue has been open for almost a year, and seems like very "low hanging fruit" which would benefit a lot of users. Is the team considering implementing it?

**Comment by danPadric** [comment]
@adamralph: _perhaps this one is kept unresolved on purpose... there should at least be one or two things that Atom is better at_ :wink:

**Comment by Deilan** [comment]
@alexandrudima Could you kindly retriage this one? A highly requested feature.

**Comment by CaiGwatkin** [comment]
In the mean time, does anyone know of any extensions that implement this feature? Doesn't look like the team are very excited to add this feature anytime soon.

**Comment by fabulousduck** [comment]
willing to send the person who adds this a snickers bar.
no bamboozles.
I mean, thats a tasty offer

**Comment by den-rain** [comment]
+1, very needed feature!

**Comment by adamralph** [comment]
Gentle reminder that lots of people are eagerly awaiting this.

**Comment by youngr2** [comment]
+1, #26646

If Render Control Characters does not show Newline characters — what control characters are rendered when enabling this setting? 

Nearly every other editor I have used with support for rendering control characters does display newlines characters when enabled.

**Comment by IbrahimTanyalcin** [comment]
This is a frequently used and highly expected feature by the users. Any developer from the VScode side that can give us a timeline of when this feature might be implemented?

**Comment by sohamkamani** [comment]
I am desperately awaiting this feature as well. 
In the meantime, I made an extension to show line endings, in case anyone wants this feature *now* : [code-eol](https://marketplace.visualstudio.com/items?itemName=sohamkamani.code-eol)

**Comment by adamralph** [comment]
@sohamkamani that's pretty cool!

Although I notice that it displays the same character `⬎` regardless of whether the EOL character is LF, CR, or CRLF. The reason I want this feature is to be able to distinguish them. If that's something you'd be willing to implement, I humbly suggest:

- LF = ↓ ([DOWNWARDS ARROW](http://www.fileformat.info/info/unicode/char/2193/index.htm))
- CR = ← ([LEFTWARDS ARROW](http://www.fileformat.info/info/unicode/char/2190/index.htm))
- CRLF = ↵ ([DOWNWARDS ARROW WITH CORNER LEFTWARDS](http://www.fileformat.info/info/unicode/char/21B5/index.htm))

These are the characters I've seen used in some other apps. The nice thing about them is that they represent the physical movement of the typewriter head when performing a line feed (LF) or carriage return (CR) and they are additive (downwards arrow + leftwards arrow = downards arrow with corner leftwards).

**Comment by luisbraganca** [comment]
+1

**Comment by sajidali** [comment]
Good work, just one more request.
@sohamkamani as @adamralph said, we need distinct rendering for each of line break characters.
I have already asked for the same in the above comments, that distinct displayed character is need to notice the mixed line ending issues.

It would be great if the official team can provide this basic feature.

**Comment by pltrant** [comment]
This missing feature is the sole reason I still have Notepad++ installed.

**Comment by sohamkamani** [comment]
Thanks @adamralph for the suggestion. Just to give you and @sajidali  an update, I have implemented and updated the [code-eol](https://marketplace.visualstudio.com/items?itemName=sohamkamani.code-eol) extension according to your suggestion. 

If there are any more suggestions, please feel free to file them under the [extensions repo](https://github.com/sohamkamani/code-eol), since we wouldn't want to pollute this thread.

Of course, what I really hope for, is for this extension to be made obsolete by vs-code having this feature natively! 😄

**Comment by adamralph** [comment]
@sohamkamani that is all kinds of awesome. Thanks!

https://twitter.com/adamralph/status/966378278020894720

**Comment by geekley** [comment]
When opening a file with mixed line endings, vscode is still showing it as `LF` or `CRLF` in the status bar.
After saving, it fixes it, but I think it should show, for example, `LF (mixed)` (maybe in red) until the file is modified.
This is important so that you know the file still has mixed EOLs, and you might want to fix that.

To reproduce: `printf "lf\ncrlf\r\ncr\rnone" > mixed-eol.txt && code mixed-eol.txt`
If you open this file in Notepad++ and enable visible EOL, you can see it has mixed endings.

By the way, this might by why code-eol isn't showing them as different EOLs.
Should I post this in a different issue, as a bug?

**Comment by geekley** [comment]
> This missing feature is the sole reason I still have Notepad++ installed.

I have a second reason: Notepad++ opens instantly.

**Comment by ericblade** [comment]
I am now working in a project where I need to have mixed line endings (some files require LF only), and this is pretty difficult to deal with in VScode. :-S  Komodo Edit has the ability to set CR/LF vs LF on specific files.  That would be a fantastic addition, for those that need it.

**Comment by tanrui8765** [comment]
In vscode version 1.36.1, it seems like the problem still exists. Is there a plan to improve this feature?

**Comment by andrecasal** [comment]
It has been almost 4 years since this feature request. I would never imagine this would be such a hard feature to implement. Are there any updates or a timeline for this?

**Comment by DougPlumley** [comment]
Any plans to implement this?

**Comment by davidc555** [comment]
Please, please address this!  Thank you

**Comment by geekley** [comment]
To everyone subscribed to this: You'll likely also want to upvote my feature request here, before it's closed by bot: https://github.com/microsoft/vscode/issues/119025

The reason most people want to see line endings is because they want to **detect and fix mixed endings**, right?
For this, I think that it's more helpful to easily know ***when*** a file has any mixed line endings (without having to look at it line by line), with option to auto-fix it on save, than to know ***where*** a file has a specific line ending (and having to manually fix it).

Knowing "where" is useful mostly in binary files, but you can currently use hex editor if you're dealing with those. It's also useful if you're working on a text file where you want mixed endings on purpose, like testing how compilers and tools deal with such files, etc., but this is an infrequent corner case.

**Comment by masaeedu** [comment]
> The reason most people want to see line endings is because they want to detect and fix mixed endings, right?

Can't speak for most people, but mixed line endings aren't always a problem that need fixing. Or at least that's not why I subscribed to this issue. I was writing a test suite that did some stuff with whitespace and wanted to be able to read the whitespace in my source code.

> Knowing "where" is useful mostly in binary files, but you can currently use hex editor if you're dealing with those. 

"Use <different editor>" is basically a solution to every feature request on this tracker, but probably not a very satisfying one to the people subscribed.

**Comment by geekley** [comment]
@masaeedu Sure, I recognize this feature is important too. In fact, the corner case I mentioned is one I'm having to deal with myself, so I'd like to have this as well.
But since the mixed lines case is closely related to this, I thought I'd notify subscribers of a better solution in fixing that particular problem (as complementary to this, not replacement). I wanted to draw attention to that feature request mostly so it doesn't die, not because it would solve the same problems as this.

**Comment by lukaszpolowczyk** [comment]
I need this feature.
I'm manually copying text from a PDF into MARKDOWN, and I need to delete unwanted new lines.
It can't be done automatically, because some new lines have to stay.

If I could see where the newline mark is, it would make my job a lot easier.
Because I have to distinguish between wrapping a line in the editor and a newline.

Disabling line wrapping is also inconvenient, because I have to move the text sideways...

**Comment by DougPlumley** [comment]
@lukaszpolowczyk out of curiosity, what's stopping you from deleting the new line characters?

You should be able to go to the end of the line or beginning of the next line and delete them.  Just because you can't see the CRLF or LF doesn't mean you can't delete it.

**Comment by lukaszpolowczyk** [comment]
@DougPlumley I'm deleting, but first I need to guess where the newline character is and where the wrapped line is.

In the PDF file, it goes something like this:

> Lorem Ipsum is simply dummy text of the printing and typesetting
 industry. Lorem Ipsum has been the industry's standard dummy
text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged.

And the editor additionally wraps some lines.

The result is that I have **two newline characters to delete**, **one wrap that I don't need to delete**, and **one newline character that should stay**.

In the editor I **cannot see** where the newline is and where the wrap is.

Do you understand now? :D

**Comment by trapper2hin** [comment]
🧐

**Comment by DougPlumley** [comment]
> @DougPlumley I'm deleting, but first I need to guess where the newline character is and where the wrapped line is.
> 
> In the PDF file, it goes something like this:
> 
> > Lorem Ipsum is simply dummy text of the printing and typesetting
> > industry. Lorem Ipsum has been the industry's standard dummy
> > text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
> > It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged.
> 
> And the editor additionally wraps some lines.
> 
> The result is that I have **two newline characters to delete**, **two wrappers that I don't need to delete**, and **one newline character that should stay**.
> 
> In the editor I **cannot see** where the newline is and where the wrap is.
> 
> Do you understand now? :D

Yes and no I suppose, because you can show line numbers and anything that wraps is going to have the same line number.

You can also select a single line feed and Control + D (Command + D on macOS) to select the next word

You can also use search in the file to find, find/replace new lines.

**Comment by StephenHodgson** [comment]
We're still waiting...

**Comment by RubenatorX** [comment]
As someone that helps new programmers learn PHP, the fact that there isn't a setting for showing newlines is really unfortunate... for example...
```php
<?php require('thing.php'); ?>
<?php
// code and whatever here
header('Location: /index.php');
```
this `header` line does not work because line 1 outputs a newline, but students do not understand that there is a newline character at the end of line 1 that is being output to the response body.

Whereas in Notepad++, the settings can make this mistake perfectly obvious:
![image](https://user-images.githubusercontent.com/5944121/231529592-d15ba87a-23ac-4c4c-83fb-2f17ec660582.png)

**Comment by Kemeros** [comment]
Would love this feature as well.

There is clearly interest for it.

Hope we get a kind soul interested in this.

Pretty please. <3

**Comment by salda** [comment]
I want it, but just for diff comparison purpose, because very often I don't see a change and usually it's because of line endings.

**Comment by jessdwitch** [comment]
We're coming up on 8 years since this was proposed. It'd be wild if this bug went a full decade without attention. And it is a bug. If my config is telling me it's rendering "all" whitespace, but there's exceptions, well then that's not "all" whitespace is it? I can't imagine it'd be super difficult to get the settings to stop lying to users. So alternative solution: change the setting to say "most", drop the "all", and just accept that truly rendering all whitespace is just beyond the technical abilities of VSC.

**Comment by thernstig** [comment]
So even Github CoPilot, with the `@vscode` chat participant, will hallunicate me this:

![Image](https://github.com/user-attachments/assets/5dcca5d4-8650-4137-ba6d-859ed1717bdc)

So this does not work unfortunately:
```json
{
    "editor.renderWhitespace": "all",
    "editor.renderControlCharacters": true
}
```

**Comment by andrecasal** [comment]
Seriously guys, it's been 9 years. Can we **please** push this feature to a higher priority so it doesn't keep falling through the cracks?

**Comment by ericblade** [comment]
I would've thought that during the creation of WSL someone at Microsoft would've had a fit about line endings but... I guess they just dealt with it lol

**Comment by thanhnguyen-luxoft** [comment]
Hi guys, I used this VS Ext to show line endings. 
https://github.com/Steditor/vsce-line-endings

**Comment by cos4ni2s** [comment]
It has been 10 years. How hard is it to implement this for VSCode maintainers?
